### PR TITLE
perf(core): make sanitization tree-shakable in Ivy mode

### PR DESCRIPTION
**Pull request by mhevery**
In VE the `Sanitizer` is always available in `BrowserModule` because the VE retrieves it using injection. 

In Ivy the injection is optional and we have instructions instead of component definition arrays. The implication of this is that in Ivy the instructions can pull in the sanitizer only when they are working with a property which is known to be unsafe. Because the Injection is optional this works even if no Sanitizer is present. So in Ivy we first use the sanitizer which is pulled in by the instruction, unless one is available through the `Injector` then we use that one instead.

This PR does few things:
1) It makes `Sanitizer` optional in Ivy.
2) It makes `DomSanitizer` tree shakable. 
3) It aligns the semantics of Ivy `Sanitizer` with that of the Ivy sanitization rules.
4) It refactors `DomSanitizer` to use same functions as Ivy sanitization for consistency.

### CODE REVIEW

File: packages/core/src/sanitization/bypass.ts [hunk 1]
```diff
@@ -6,61 +8,113 @@
...
 
-const BRAND = '__SANITIZER_TRUSTED_BRAND__';
-
 export const enum BypassType {
-  Url = 'Url',
-  Html = 'Html',
-  ResourceUrl = 'ResourceUrl',
+  Url = 'URL',
+  Html = 'HTML',
+  ResourceUrl = 'ResourceURL',
```
Comment (IgorMinar) [inline packages/core/src/sanitization/bypass.ts]: were these changes necessary or are they just stylistic?

File: packages/platform-browser/src/browser.ts [hunk 2]
```diff
@@ -37,9 +37,13 @@ export const INTERNAL_BROWSER_PLATFORM_PROVIDERS: StaticProvider[] = [
  */
 export const BROWSER_SANITIZATION_PROVIDERS: StaticProvider[] = [
   {provide: Sanitizer, useExisting: DomSanitizer},
-  {provide: DomSanitizer, useClass: DomSanitizerImpl, deps: [DOCUMENT]},
 ];
 
+export const BROWSER_SANITIZATION_PROVIDERS__POST_R3__ = [];
+export const BROWSER_SANITIZATION_PROVIDERS__PRE_R3__ = BROWSER_SANITIZATION_PROVIDERS;
```
Comment (IgorMinar) [inline packages/platform-browser/src/browser.ts]: is this right? I thought the right way to do this was to assign `{provide: Sanitizer, useExisting: DomSanitizer}` to `BROWSER_SANITIZATION_PROVIDERS__PRE_R3__` and then assign `BROWSER_SANITIZATION_PROVIDERS__PRE_R3__` to `BROWSER_SANITIZATION_PROVIDERS`

Comment (IgorMinar) [reply packages/platform-browser/src/browser.ts]: you seem to be doing it backwards

File: packages/platform-browser/src/browser.ts [hunk 3]
```diff
@@ -37,9 +37,13 @@ export const INTERNAL_BROWSER_PLATFORM_PROVIDERS: StaticProvider[] = [
...
+
+const _BROWSER_SANITIZATION_PROVIDERS_INTERNAL = BROWSER_SANITIZATION_PROVIDERS__PRE_R3__;
```
Comment (IgorMinar) [inline packages/platform-browser/src/browser.ts]: if you do it the way I suggested there is no need for this const

File: packages/platform-browser/src/private_export.ts [hunk 4]
```diff
@@ -20,3 +20,9 @@ export {HammerGesturesPlugin as ɵHammerGesturesPlugin} from './dom/events/hamme
 export {KeyEventsPlugin as ɵKeyEventsPlugin} from './dom/events/key_events';
 export {DomSharedStylesHost as ɵDomSharedStylesHost, SharedStylesHost as ɵSharedStylesHost} from './dom/shared_styles_host';
 export {DomSanitizerImpl as ɵDomSanitizerImpl} from './security/dom_sanitization_service';
+
+// we reexport these symbols just so that they are retained during the dead code elimination
+// performed by rollup while it's creating fesm files.
+//
+// no code actually imports these symbols from the @angular/platform-browser entry point
+export {BROWSER_SANITIZATION_PROVIDERS__POST_R3__} from './browser';
```
Comment (IgorMinar) [inline packages/platform-browser/src/private_export.ts]: won't this still make it part of the public api unless you prefix it with ɵ?

File: packages/platform-browser/test/browser/bootstrap_spec.ts [hunk 5]
```diff
@@ -189,6 +189,19 @@ function bootstrap(
...
 
+    it('should retrieve sanitizer', inject([Injector], (injector: Injector) => {
+         const sanitizer: Sanitizer|null = injector.get(Sanitizer, null);
+         if (ivyEnabled) {
+           // In Ivy we don't want to have sanitizer in DI. We use DI only to overwrite the
+           // sanitizer, but not for default one. The default one is pulled in by the Ivy
+           // instructions as needed.
+           expect(sanitizer).toBe(null);
+         } else {
+           // In VE we always need to have sanitize available.
```
Comment (IgorMinar) [inline packages/platform-browser/test/browser/bootstrap_spec.ts]:
```suggestion
           // In VE we always need to have Sanitizer available.
```

**Review by IgorMinar — COMMENTED** [review]
I need to review this more carefully to have good feedback, so I'm just posting initial observations.

Three more things:
- Can you please write up a good commit message describing the change and providing enough context for all the changes you made?
- Can you take a look at this older PR that Martin has been trying to land for a while: https://github.com/angular/angular/pull/12310 and see if it makes sense to incorporate it into this change?
- We will likely need Rafael or someone else from the security team to take a look at this as well as the changes you made are nontrivial.

File: packages/core/src/sanitization/bypass.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/sanitization/bypass.ts]: Necessary to align the Ivy sanitization with `DomSanitizer`.

File: packages/platform-browser/src/browser.ts (hunk 2 above)
Comment (mhevery) [reply packages/platform-browser/src/browser.ts]: Yeah, sorry fixed.

File: packages/platform-browser/src/browser.ts (hunk 3 above)
Comment (mhevery) [reply packages/platform-browser/src/browser.ts]: removed

File: packages/platform-browser/src/private_export.ts (hunk 4 above)
Comment (mhevery) [reply packages/platform-browser/src/private_export.ts]: correct, fixed.

File: packages/platform-browser/test/browser/bootstrap_spec.ts (hunk 5 above)
Comment (mhevery) [reply packages/platform-browser/test/browser/bootstrap_spec.ts]: fixed

**Review by IgorMinar — COMMENTED** [review]
a few more comments / suggestions

File: packages/core/src/sanitization/bypass.ts [hunk 6]
```diff
@@ -6,61 +6,120 @@
...
 
 /**
- * A branded trusted string used with sanitization.
+ * Marker interface for a value that's safe to use in a particular context.
  *
- * See: {@link TrustedHtmlString}, {@link TrustedResourceUrlString}, {@link TrustedScriptString},
- * {@link TrustedStyleString}, {@link TrustedUrlString}
+ * @publicApi
  */
-export interface TrustedString extends String { [BRAND]: BypassType; }
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/sanitization/bypass.ts]: I did not remove anything, I just moved it from `platform-browser` to the `core`. We could add additional branding if you would like, but that is not what we had in the past.

File: packages/core/src/sanitization/bypass.ts [hunk 7]
```diff
@@ -6,61 +6,120 @@
...
+    value: any, type: BypassType.ResourceUrl): value is SafeResourceUrl;
+export function allowSanitizationBypassAndThrow(
+    value: any, type: BypassType.Script): value is SafeScript;
+export function allowSanitizationBypassAndThrow(
+    value: any, type: BypassType.Style): value is SafeStyle;
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType.Url): value is SafeUrl;
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean;
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean {
+  const actualType = getSanitizationBypassType(value);
+  if (actualType != null && actualType !== type) {
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/sanitization/bypass.ts]: No, I actually want this to be sensitive to `null` and `undefined`

File: packages/core/src/sanitization/bypass.ts [hunk 8]
```diff
@@ -6,61 +6,120 @@
...
+    // Allow ResourceURLs in URL contexts, they are strictly more trusted.
+    if (actualType == BypassType.ResourceUrl && type == BypassType.Url) return true;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/sanitization/bypass.ts]: fixed

File: packages/platform-browser/src/security/dom_sanitization_service.ts [hunk 9]
```diff
@@ -154,78 +155,39 @@ export class DomSanitizerImpl extends DomSanitizer {
...
         return _sanitizeHtml(this._doc, String(value));
       case SecurityContext.STYLE:
-        if (value instanceof SafeStyleImpl) return value.changingThisBreaksApplicationSecurity;
-        this.checkNotSafeValue(value, 'Style');
+        if (allowSanitizationBypassAndThrow(value, BypassType.Style)) return unwrapSafeValue(value);
         return _sanitizeStyle(value as string);
       case SecurityContext.SCRIPT:
-        if (value instanceof SafeScriptImpl) return value.changingThisBreaksApplicationSecurity;
-        this.checkNotSafeValue(value, 'Script');
+        if (allowSanitizationBypassAndThrow(value, BypassType.Script))
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/platform-browser/src/security/dom_sanitization_service.ts]: fixed

File: packages/core/src/sanitization/sanitization.ts [hunk 10]
```diff
@@ -38,8 +38,8 @@ export function ɵɵsanitizeHtml(unsafeHtml: any): string {
   if (sanitizer) {
     return sanitizer.sanitize(SecurityContext.HTML, unsafeHtml) || '';
   }
-  if (allowSanitizationBypass(unsafeHtml, BypassType.Html)) {
-    return unsafeHtml.toString();
+  if (allowSanitizationBypassAndThrow(unsafeHtml, BypassType.Html)) {
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/sanitization/sanitization.ts]: the function throws if the sanitization is not valid. Can you suggest a different name?

File: packages/platform-browser/src/security/dom_sanitization_service.ts [hunk 11]
```diff
@@ -154,78 +155,39 @@ export class DomSanitizerImpl extends DomSanitizer {
...
+          return unwrapSafeValue(value);
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/platform-browser/src/security/dom_sanitization_service.ts]: fixed

File: packages/core/src/sanitization/bypass.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/core/src/sanitization/bypass.ts]: Oh I see what you are doing... rather than use branded types you have switched to using real classes (e.g. `SafeValueImpl` and descendants). 

But in that case I don't see any value in having both real implementations and interfaces... especially since the interfaces are all semantically identically now...

```
const x: SafeValue = new SafeHtmlImpl('some string');
const y: SafeScript = x;  // No compile time error...
```

Why not just make them all classes and ditch the interfaces?

File: packages/core/src/sanitization/bypass.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/core/src/sanitization/bypass.ts]: At least in the compiler we have been trying to be more explicit about these things, so if we wanted to compare both `null` and `undefined` then we would write it out explicitly using `===`. This would ensure people like @IgorMinar don't mistake this sort of comparison as an error. I think the minimizer might be able to collapse such comparisons back to `==` at bundling time so there is no benefit in using it in the source.

Alternatively, I think a comment explaining why `!=` is being used rather than` `!==` would help.

File: packages/core/src/sanitization/bypass.ts [hunk 12]
```diff
@@ -6,61 +6,120 @@
...
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean;
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean {
+  const actualType = getSanitizationBypassType(value);
+  if (actualType != null && actualType !== type) {
+    // Allow ResourceURLs in URL contexts, they are strictly more trusted.
+    if (actualType === BypassType.ResourceUrl && type === BypassType.Url) return true;
+    throw new Error(
+        `Required a safe ${type}, got a ${actualType} (see http://g.co/ng/security#xss)`);
+  }
+  return actualType === type;
```
Comment (petebacondarwin) [inline packages/core/src/sanitization/bypass.ts]: So I think this function will:

* return `true` if
  * `actualType === type`
  * or `actualType` is `ResourceUrl` and `type` is `Url`
* return `false`: if
  * `actualType === null` and `type !== null`
  * or `actualType === undefined` and `type !== undefined`
* throw otherwise

Is that correct? It is not so obvious to glean that from reading the code.

File: packages/core/src/sanitization/sanitization.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/core/src/sanitization/sanitization.ts]: Regarding the name, I feel that the problem is that this function is trying to do two different things:

1) check whether we can bypass the sanitizer for these types
2) check whether sanitization is allowed at all for these types

It seems to me that point 2) should be a check that the sanitizer should be doing e.g. in `_sanitizeHtml` and friends.

File: packages/core/test/render3/instructions_spec.ts [hunk 13]
```diff
@@ -161,33 +161,13 @@ describe('instructions', () => {
...
-      expect(sanitizerInterceptor.lastValue !).toEqual('apple');
-      sanitizerInterceptor.lastValue = null;
-
-      t.update(() => {
-        ɵɵstyleSanitizer(sanitizerInterceptor.getStyleSanitizer());
-        ɵɵstyleProp('background-image', bypassSanitizationTrustStyle('apple'));
-        ɵɵstylingApply();
-      });
-      expect(sanitizerInterceptor.lastValue).toEqual(null);
-    });
```
Comment (petebacondarwin) [inline packages/core/test/render3/instructions_spec.ts]: Is this test no longer applicable? If it is then should it appear somewhere in the new code?

**Review by petebacondarwin — COMMENTED** [review]
I came to review for `docs-infra` but then found that you had removed all the changes in `aio` :-)

File: packages/core/test/render3/instructions_spec.ts (hunk 13 above)
Comment (mhevery) [reply packages/core/test/render3/instructions_spec.ts]: yes, this test is invalid, and has been removed.

File: packages/core/src/sanitization/bypass.ts (hunk 12 above)
Comment (mhevery) [reply packages/core/src/sanitization/bypass.ts]: Do you have suggestions as to make it more clear?

**Comment by mhevery** [comment]
@petebacondarwin The type confusion comes from the fact that I am moving code between packages. There is no change to public API. Yes your suggestions would be a great improvement but it would be breaking change as it would make it more strict and so it should be done in separate PR.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:263152058:BASE:263152086:1565713637340:c8799d)

File: packages/core/src/sanitization/bypass.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (IgorMinar) [reply packages/core/src/sanitization/bypass.ts]: That would be a breaking change.

**Review by IgorMinar — APPROVED** [review]
Lgtm, but please resolve Pete's comments.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:263152058:BASE:263219519:1565732748631:fdd5ab75)

**Comment by mhevery** [comment]
MERGE-ASSISTANCE: [Unrelated failure](http://test/OCL:263152058:BASE:263219519:1565732748631:fdd5ab75) (Unable to force green state)

**Comment by AndrewKushnir** [comment]
The changes in this PR [break some g3 targets (Guitar tests) with Ivy](https://sponge.corp.google.com/invocation?id=df9be109-85ee-407c-b4bd-79de0970b695), adding "Blocked" label for now, so we can investigate problems before merging this PR.

**Comment by AndrewKushnir** [comment]
Also started a [Global TAP Presubmit](https://test.corp.google.com/ui#id=OCL:263275270:BASE:263275368:1565755463975:e687484e) to see if there are other affected targets.

**Comment by AndrewKushnir** [comment]
- [VE Presubmit](http://test/OCL:263458988:BASE:263459016:1565826584830:2b928306)
- [Ivy Presubmit](http://test/OCL:263458988:BASE:263459193:1565826644784:1e19f193)
